### ISSUE TITLE
merge hoodie.remote into hoodie.store

### ISSUE THREAD

**Issue by gr2m** [hoodiehq/hoodie#91]
I see people get confused by hoodie.store and hoodie.remote. Especially when it comes to event listeners. The 2 modules are there because of historic reasons. But after thinking trough it again I think it's exposing the technical architecture behind Hoodie, that indeed there is a local store in the browser and a central store on the server. And we should hide away that if we can.

So, for reference, that's the API that we have at the moment:

``` js
// STORE
hoodie.store.add(type, attributes )
hoodie.store.update( type, id, update )
hoodie.store.find( type, id )
hoodie.store.findAll()
hoodie.store.remove( type, id )
hoodie.store.removeAll()

hoodie.store.on( 'add',    function( newObject, options) { } )
hoodie.store.on( 'update', function( updatedObject, options) { } )
hoodie.store.on( 'remove', function( removedObject, options) { } )
hoodie.store.on( 'change', function( eventName, changedObject, options) { } )

// REMOTE
// Remote has all the store methods like `.find`, plus:
hoodie.remote.connect()
hoodie.remote.disconnect()
hoodie.remote.pull()
hoodie.remote.push()
hoodie.remote.sync()

hoodie.remote.on( 'add',    function( newObject, options) { } )
hoodie.remote.on( 'update', function( updatedObject, options) { } )
hoodie.remote.on( 'remove', function( removedObject, options) { } )
hoodie.remote.on( 'change', function( eventName, changedObject, options) { } )
```

One already can see that they are quite similar. And through the `options` parameter in the `hoodie.store.on` callback, one can differentiate between wether a change comes from remote or from local.

So I'd suggest to get rid of `hoodie.remote` entirely, and add the methods: `connect`, `disconnect`, `pull`, `push`, `sync` to `hoodie.store`.

What you think?

**Comment by Acconut** [comment]
+1

**Comment by svnlto** [comment]
:+1:

**Comment by gr2m** [comment]
The complex thing here is that hoodie.store would "inherit" from both `Hoodie.Store` and `Hoodie.Remote`. 

`connect`, `disconnect`, `pull`, `push`, `sync` would be used from `Hoodie.Remote`, but not the store methods, like `find`, `add`, `remove` etc.

The whole inheritance thing might not work here, or only with ugly workarounds. Now that we left CoffeeScript behind us, we could take this issue as an example on how the hoodie.js architecture should be, lets put together some dreamcode?

I'd love to hear @dilvie's thoughts on this.

**Comment by ericelliott** [comment]
I think that remote should be wrapped by store, and hidden from the users - treated like an implementation detail.

Users should not be bothered about whether a data store is local or remote. Regardless, data I/O should be treated as asynchronous and error-prone. If you do that, you can completely abstract the store API from networking and synchronization concerns. _"Program to an interface, not an implementation."_

Further, I think it would be OK to keep remote around in case programmers need a lower level abstraction for working with remote data -- but I don't think it should have any of store's methods.

Think about it this way: if you have both store and remote hanging around, and one is just a slightly more capable version of the other, what's the point of keeping them both? It's just a lot of duplicated API surface area, only they will have slightly divergent properties, which is a potential source of confusion to users. This is a famous problem in OO design - **the unexpected divergence problem**.

In other words, use remote inside store, but don't expose remote's API from store. Instead, handle synchronization internally. Users who need more control can build on top of the raw materials supplied by either remote, or store, or both.

All of that said, there are some much deeper problems with the OO design underlying a lot of hoodie. I'll get to a much more detailed critique eventually, but if you haven't yet, you can start by watching [Classical Inheritance is Obsolete: How to Think in Prototypal OO](https://vimeo.com/69255635).

**Comment by gr2m** [comment]
Thanks Eric! Actually your talk on Pototypal OO is why I contacted you in the first place. Besides your expertise, I really appreciate the way you argue and your sense for both the user experience of the API as well as the architecture behind the curtain. And @janl just seconded that ;-)

To give you some Background, here are Hoodie Modules that depend on Store
1. [Hoodie.Store](https://github.com/hoodiehq/hoodie.js/blob/master/src/core/store.js) defines the generic store api and has some validations build in, it also implements methods like `update` that are the same no matter of the I/O, to prevent code duplication. `Hoodie.Store` is used internally only.
2. [Hoodie.LocalStore](https://github.com/hoodiehq/hoodie.js/blob/master/src/core/local_store.js) is the module that is exposed as `hoodie.store` to the user. It inherits from `Hoodie.Store` and implements the I/O using localStorage, for offline data storage / caching. It also listens to new objects coming from `hoodie.remote` and stores them as well.
3. [Hoodie.Remote](https://github.com/hoodiehq/hoodie.js/blob/master/src/core/remote.js). `Hoodie.Remote` is not exposed directly, it gets initialized when using `hoodie.open( storeName )`. It implements the I/O using REST requests against CouchDBs API. So `hoodie.open( 'public_data' ).findAll().then( display )` would load all objects from the database `public_data` using `GET /_api/public_data/_all_docs` and pass the result as array to the `display` function.
4. [Hoodie.AccountRemote](https://github.com/hoodiehq/hoodie.js/blob/master/src/core/account_remote.js) is based on `Hoodie.Remote` and is currently exposed as `hoodie.remote`. It's tied to the user's database and has some added functionality, like pushing changed objects from `hoodie.store` and triggering events on updates from remote, so `hoodie.store` can save these.
5. [Hoodie.ShareInstance](https://github.com/hoodiehq/hoodie.js/blob/master/src/extensions/share_instance.js) is based on `Hoodie.Remote` as well, tied to a share database and extending the remote store API with methods like `grantReadAccess`, `grantWriteAccess` or `subscribe`. It also extends the `hoodie.store` API, so you can do things like `hoodie.store.findAll('task').shareAt( shareId )`

Hope that's not too confusing. So although we wouldn't expose the `Hoodie.Remote` directly, I'd still like to allow developers to build plugins by somehow inheriting from its functionality, as it will be very common to build plugins that will have their own databases and provide some nice API to work with them. There are also [Hoodie.User](https://github.com/hoodiehq/hoodie.js/blob/master/src/extensions/user.js) and [Hoodie.Global](https://github.com/hoodiehq/hoodie.js/blob/master/src/extensions/global.js) that don't extend `Hoodie.Remote` directly, but just return an instance tied to a specific database.

**Comment by ericelliott** [comment]
Time to test the thick skin! =)

The whole approach is very class-like, and a bit awkward, from an API user's perspective. Here are some changes to the over-all OO style that I would make:

## Make it more prototypal
1. There should be no constructors, and no need for capital letters to distinguish constructors, or `new`. This solves [a wide range](http://ericleads.com/2013/01/javascript-constructor-functions-vs-factory-functions/) of [problems](http://ericleads.com/2012/09/stop-using-constructor-functions-in-javascript/).
2. All interfaces should be defined as exemplar prototypes.

## Handling of internal implementation details

A good API is self documenting. It only exposes the things that the user should be concerned with. A good API is also focused. Every method you expose should relate plainly to a primary purpose. Often, internal methods are utilities that don't fit with the rest of the API surface area, and that in itself is a good reason not to expose them.

The underscore convention is problematic. By exposing properties, you make an implied contract that those properties will always be there, and always behave in a manner consistent with their current behavior. Invariably, when you give people a footgun (by opening up your internals for them to muck with), somebody gets shot in the foot. I know, I know, you're thinking, "but the underscore says that you shouldn't mess with it!"

That's ridiculous. Trust me. Newbies don't know the meaning of the underscore, and advanced developers think, "that's not for me, I know what I'm doing." Either way, your interface contract gets violated, and you either support it for bugwards compatibility, or you break a bunch of userland code when you need to make changes.

Got some private stuff? Hide it in a closure. There's no excuse to do otherwise.

While we're at it, using underscores for your closure variables doesn't make sense, either. Instance vars will be prefixed with `this.` -- you don't need anything else to make the distinction. Underscores reek of hungarian notation -- there's good reason that fell out of favor.

## Implicit globals

Hoodie is an implicit global that modules are reaching out and trying to grab directly. You should be using one of the commonly available module patterns, instead. Import your promises module and your Hoodie errors explicitly. I don't care if you use AMD or Node style (but I strongly prefer Node style). Just don't do what you're currently doing. You need to make all your dependencies explicit so that it's easier to make changes without breaking things.

## Unused private

  Store.prototype._now is unused. If it's used anywhere else in Hoodie, you're breaking your own private contract.

## Now for some code

Sort that all out, and it starts to look something [like this](https://gist.github.com/dilvie/5928512) -- but that's not quite doing everything that's needed.

You also need to be able to easily compose this functionality into other objects. This is where Stampit would come in really handy. Just require stampit and export stampit(prototype), and now you can `.compose()` store with other functionality -- last-in overrides.

hoodie/localStore becomes:

``` js
var stampit = require('stampit'),
  store = require('hoodie/store'),

  overrides = { /* stuff to override */};

module.exports = stampit.compose(store, overrides);
```

Stampit differs from $.extend() in that it has special handling for three types of prototypes - delegates (the normal prototype object you think of when you're using the constructor function prototype property), clones and mixins (which I recommend for state), and functional prototypes (for instance-safe data privacy and encapsulation).

The stampit version looks something [like this](https://gist.github.com/dilvie/5928583).

Going further - store could define a public API, and localStore could define private implementation details. Instead of letting localStore override functions willy-nilly, you can set up hooks that store calls. In other words, store could be a implemented as a facade. You can even have those hooks take multiple callbacks, so that when a method is called on store, it also gets called on both localStore and remote. While I've rolled that pattern from scratch many times, I haven't ever abstracted it into a library before. That might be a useful exercise.

**Comment by gr2m** [comment]
> Store.prototype._now is unused. If it's used anywhere else in Hoodie, you're breaking your own private contract

I just needed a method that I can stub for testing. Would be great to get rid of it.

I'm still digesting the rest ;-) It all sounds great, just want to give it some time to sink. Are there any downsides that you can think of when using stampit for hoodie.js, especially when it comes to developers of hoodie.js plugins?

**Comment by ericelliott** [comment]
> Are there any downsides that you can think of when using stampit for hoodie.js, especially when it comes to developers of hoodie.js plugins?

The downside is that what I'm suggesting is a fundamental paradigm shift in the way that you treat OO design in Hoodie, and some of that might entail some breaking changes. Particularly where it comes to making everything more modular, removing all requirements for `new`, and so on. I think making the switch would require a major version change, and plenty of signaling to the user community that the change is coming, and why it's important to Hoodie, and the projects that rely on Hoodie. Basically, it'll force a lot of users to make changes in their own code, but it will also make Hoodie, and the userland code that relies on Hoodie a lot more flexible.

You may want to consider a grace period where you port serious bug fixes to the old version of Hoodie, in a parallel branch or repo.

Sounds like a lot of work. So the question you should be asking yourself is, "is this really worth the effort and the impact on the user community?"

I think it is.

I think that the idea of Hoodie extensions as it stands is particularly dangerous, and could entice an extension user community like the one around jQuery, only with hoodie, you're much closer to the architectural foundations of an application (whereas jQuery extensions are more like fancy window curtains). Hoodie's extension mechanism encourages a monolithic codebase built on top of hoodie -- and that could go sideways very easily.

One obvious issue with the extension ecosystem is name collision and property conflicts. When everything attaches to hoodie, it's similar to introducing another global playground for module and property names, and that can cause all sorts of state related bugs, timing errors, etc... Contrast that to modules, which have independent spaces for names, completely removing the threat of methodname and property collisions.

It isn't just the user community you should be worried about. You already have a bunch of core extensions, and I can think of many more that would be tempting to add to Hoodie going forward, so even if you didn't expose the extension mechanism to users, I could see Hoodie itself turning into a monolithic framework that could potentially pose problems for users down the road (for many of the reasons I mentioned in my talk).

Making hoodie more modular means that users can select just the functionality they need, without inheriting all of hoodie's jungle and gorillas. ;)

**Comment by gr2m** [comment]
I don't worry too much about API changes at this point of the project, if we decide that they are the right thing to do. I'd suggest we move on in the repo at https://github.com/gr2m/hoodie-stampit-dreamcode.js to see where it goes.

I'd like to get a good understanding of how you think Hoodie should be build, without thinking of any constraints at this point. Once we got that, I'd think about if it's needed and worth to change the API, but let's don't think about it for the moment.

Lemme close this ticket, it kinda went off topic. I'll open a new issue to merge `hoodie.remote` into `hoodie.store`. And for questions  / discussions on hoodie-stampit, let use issues and pull requests at https://github.com/gr2m/hoodie-stampit-dreamcode.js
